Add useRawValue option to avoid encoding/decoding

loganrosen opened the pull request:
This commit adds a new `encodeValue` option to ember-cookies. I made it `true` by default to match current behavior. When it is set to `false`, it bypasses encoding the value of the cookie, in case that behavior is not desired by the user.

I personally have a scenario where I don't want the cookie to be encoded automatically, hence making this change.

marcoow commented:
> I personally have a scenario where I don't want the cookie to be encoded automatically, hence making this change.

Can you elaborate on that use case a bit?

loganrosen commented:
My use case requires that there be a vanilla `=` in the cookie value, since the value serves as its own key-value pair. The code I'm interacting with splits on an `=` and that character only. It'll be very difficult to change it to, for example, split on `%3D` instead (I don't own the code in question).

Regardless, though, I don't think `encodeUriComponent()` should be forced on the cookie value if the user doesn't want it, since there [are a bunch of characters](https://stackoverflow.com/a/1969339/789569), including `=`, that are technically valid in a cookie value but are encoded by that function.

My workaround for now is to call the `_writeDocumentCookie()` method directly, and that's not ideal (since it's not part of your library's public API). I'd really like to continue using your addon since it's definitely the best cookie one for Ember.

marcoow commented:
I see. You'd have to support opting out of decoding to the `read` method as well so you'll be able to read the correct value for an unencoded cookie.

marcoow commented on tests/unit/services/cookies-test.js:
```diff
@@ -187,6 +187,13 @@ describe('CookiesService', function() {
+        this.subject().write(COOKIE_NAME, value, { encodeValue: false });
+
+        expect(document.cookie).to.include(`${COOKIE_NAME}=${value}`);
+      });
```
A similar test would have to be added to the section that tests correct behavior in FastBoot below.

marcoow commented on addon/services/cookies.js:
```diff
@@ -67,10 +67,16 @@ export default Service.extend({
     assert("Cookies cannot be set as signed as signed cookies would not be modifyable in the browser as it has no knowledge of the express server's signing key!", !options.signed);
     assert('Cookies cannot be set with both maxAge and an explicit expiration time!', isEmpty(options.expires) || isEmpty(options.maxAge));
-    value = this._encodeValue(value);
+    assert('The encodeValue option must be a boolean!', typeof options.encodeValue === 'boolean');
```
I don't think we need that - if someone want to set a truthy value I'd say that's fine.

marcoow commented on addon/services/cookies.js:
```diff
@@ -67,10 +67,16 @@ export default Service.extend({
   write(name, value, options = {}) {
+    if (isEmpty(options.encodeValue)) {
+      options.encodeValue = true;
+    }
```
Let's use `assign` for dealing with the defaults:

```js
const DEFAULTS = { // this should probably be defined somewhere outside of this method.
  color : 'red',
};
options = assign({}, defaults, options || {});
```

loganrosen commented:
@marcoow Have you gotten a chance to look at my latest PR yet? I believe that I addressed all of your concerns. Also renamed the option to be clearer and more applicable to both writing and reading cookies.

marcoow commented on addon/services/cookies.js:
```diff
@@ -50,7 +51,10 @@ export default Service.extend({
-  read(name) {
+  read(name, options = {}) {
+    options = assign({}, DEFAULTS, options || {});
+    assert('Domain, Expires, Max-Age, and Path options cannot be set when reading cookies', isEmpty(options.domain) && isEmpty(options.expires) && isEmpty(options.maxAge) && isEmpty(options.path));
```
Can you add tests for this (see e.g. https://github.com/loganrosen/ember-cookies/blob/9392c6d2504835c1bce10ff1503f63e241672972/tests/unit/services/cookies-test.js#L18)?

marcoow commented on tests/unit/services/cookies-test.js:
```diff
@@ -87,6 +87,23 @@ describe('CookiesService', function() {
+      it("doesn't decode the value when useRawValue is true", function() {
+        let value = '!"§$%&/()=?"';
+        let encodedValue = encodeURIComponent(value);
+        document.cookie = `${COOKIE_NAME}=${encodedValue}`;
```
I guess it would be easier here to simple use the un-encoded value `document.cookie = `${COOKIE_NAME}=${value}`;` and change the assertion to `expect(afterRoundtrip).to.eq(value);`?

loganrosen replied:
Would that be a proper test for this scenario, though? Doesn't the set value need to be URI-encoded in order to verify that it's not URI-decoding it when `raw` is `true`?

marcoow replied:
Yep, you're right of course 🤦‍♂️

marcoow commented on tests/unit/services/cookies-test.js:
```diff
@@ -358,6 +382,20 @@ describe('CookiesService', function() {
 
+      it('URI-component-decodes the value', function() {
+        let value = '!"§$%&/()=?"';
+        this.subject().write(COOKIE_NAME, value);
```
I think you'd explicitly want to write the encoded value here - similar to https://github.com/simplabs/ember-cookies/pull/134/files#diff-91737467d778614edb8e37696092b40bR92

loganrosen replied:
Why is that necessary if `write()` is already URI-encoding the value by default? Are you saying that I should pass `encodeURIComponent(value)` to the function just to ensure that it's encoded?

marcoow replied:
I'd say you don't want to go through `write` at all here as that makes the test depend on the correctness of `write` and also a bit harder to understand. You can simply use `document.cookie` directly as in https://github.com/simplabs/ember-cookies/pull/134/files/9392c6d2504835c1bce10ff1503f63e241672972#diff-91737467d778614edb8e37696092b40bR101

loganrosen replied:
I'm using `document.cookie` directly in the regular (non-Fastboot) context. But all of the other unit tests for reading Fastboot cookies involve calling the `write()` function to write them. Should those be changed too? 🤔

marcoow replied:
Ah sorry, this is a FastBoot test of course 🤦‍♂️ Yes, I do thing we should change those tests to set the header directly instead of relying on `write` but that's something for another PR.

marcoow commented on addon/services/cookies.js:
```diff
@@ -6,7 +6,8 @@ import { A } from '@ember/array';
 import { merge } from '@ember/polyfills';
-const { keys } = Object;
+const { keys, assign } = Object;
+const DEFAULTS = { useRawValue: false };
```
I'd rename this to just `raw`.

marcoow commented on README.md:
```diff
@@ -47,10 +47,10 @@ export default Ember.Controller.extend({
-* `read(name)`: reads the cookie with the given name, returns its value as a
-  `String`.
+* `read(name, options = {})`: reads the cookie with the given name, returns its value as a
+  `String`; options can be used to set `useRawValue` (boolean).
```
we should add something like _"…set `raw` which disables (boolean) URL-decoding"_

marcoow commented on README.md:
```diff
@@ -47,10 +47,10 @@ export default Ember.Controller.extend({
+  `String`; options can be used to set `useRawValue` (boolean).
 * `write(name, value, options = {})`: writes a cookie with the given name and
-  value; options can be used to set `domain`, `expires` (Date), `maxAge` (time in seconds), `path` and `secure`.
+  value; options can be used to set `domain`, `expires` (Date), `maxAge` (time in seconds), `path`, `secure`, and `useRawValue` (boolean).
```
we should add something like _"…and `raw` which disables (boolean) URL-encoding the value"_

loganrosen commented:
@marcoow I believe I've finally addressed all of your comments. Please take a look and let me know if there's anything further to change. :slightly_smiling_face:

marcoow commented on README.md:
```diff
@@ -47,10 +47,11 @@ export default Ember.Controller.extend({
-* `read(name)`: reads the cookie with the given name, returns its value as a
-  `String`.
+* `read(name, options = {})`: reads the cookie with the given name, returns its value as a
+  `String`; options can be used to set `raw` (boolean, disables URL-encoding the value).
```
This should be _“…disables URL-decoding…“_

loganrosen commented:
@marcoow Done. 🙂

marcoow commented:
@loganrosen: I think you changed the wrong one - it should be _"disables **de**coding"_ for `read` and _"disables **en**coding"_ for `write` I think? 😬

loganrosen commented:
@marcoow I did. 🤦‍♂️ Fixed!
